src/devices/laserFromExternalPort/laserFromExternalPort.cpp

```diff
@@ -0,0 +1,216 @@
+ */
+
+#define _USE_MATH_DEFINES
+
+#include "laserFromExternalPort.h"
+
+#include <yarp/os/Time.h>
+#include <yarp/os/Log.h>
+#include <yarp/os/LogStream.h>
+#include <yarp/os/ResourceFinder.h>
+
+#include <cmath>
+#include <cstring>
+#include <cstdlib>
+#include <iostream>
+#include <limits>
+#include <mutex>
+
+using namespace std;
+
+#ifndef DEG2RAD
+#define DEG2RAD M_PI/180.0
+#endif
+
+InputPortProcessor::InputPortProcessor()
+{
+}
+
+void InputPortProcessor::onRead(yarp::dev::LaserScan2D& b)
+{
+    mutex.lock();
+        lastScan = b;
+        getEnvelope(lastStamp);
+    mutex.unlock();
+}
+
+inline void InputPortProcessor::getLast(yarp::dev::LaserScan2D& data, Stamp& stmp)
+{
+    mutex.lock();
+        data = lastScan;
+        stmp = lastStamp;
+    mutex.unlock();
+}
+
+//-------------------------------------------------------------------------------------
+
+bool LaserFromExternalPort::open(yarp::os::Searchable& config)
+{
+    Property subConfig;
+    m_info = "LaserFromExternalPort device";
+    m_port_name = "/laserFromExternalPort:i";
+
+#ifdef LASER_DEBUG
+    yDebug("%s\n", config.toString().c_str());
+#endif
+
+    if (this->parse(config) == false)
+    {
+        yError() << "LaserFromExternalPort: error parsing parameters";
+        return false;
+    }
+
+    yarp::os::Searchable& general_config = config.findGroup("SENSOR");
+    if (general_config.check("input_port_name"))
+       {m_port_name = general_config.find("input_port_name").asString();} //this parameter is optional
+
+    m_input_port.useCallback();
+    m_input_port.open(m_port_name);
+
+    PeriodicThread::start();
+
+    yInfo("LaserFromExternalPort: Sensor ready");
+    return true;
+}
+
+bool LaserFromExternalPort::close()
+{
+    PeriodicThread::stop();
+
+    m_input_port.close();
+
+    yInfo() << "LaserFromExternalPort closed";
+    return true;
+}
+
+
+
+bool LaserFromExternalPort::setDistanceRange(double min, double max)
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+    m_min_distance = min;
+    m_max_distance = max;
+    return true;
+}
+
+bool LaserFromExternalPort::setScanLimits(double min, double max)
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+    yWarning("setScanLimits not yet implemented");
+    return true;
+}
+
+
+
+bool LaserFromExternalPort::setHorizontalResolution(double step)
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+    yWarning("setHorizontalResolution not yet implemented");
+    return true;
+}
+
+bool LaserFromExternalPort::setScanRate(double rate)
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+    yWarning("setScanRate not yet implemented");
+    return false;
+}
+
+
+
+bool LaserFromExternalPort::threadInit()
+{
+#ifdef LASER_DEBUG
+    yDebug("LaserFromExternalPort:: thread initialising...\n");
+    yDebug("... done!\n");
+#endif
+
+    return true;
+}
+
+void LaserFromExternalPort::run()
+{
+#ifdef DEBUG_TIMING
+    double t1 = yarp::os::Time::now();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 't1' is assigned a value that is never used.](https://app.codacy.com/manual/robotology/yarp/pullRequest?prid=5034547)